src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteEmbeddedLanguage/Infrastructure/RoutePatternParametersDetector.cs

```diff
@@ -0,0 +1,131 @@
+        };
+
+        foreach (var child in childSymbols)
+        {
+            if (child.HasAttribute("Microsoft.AspNetCore.Http.AsParametersAttribute", semanticModel))
+            {
+                resolvedParameterSymbols.AddRange(ResolvedParameters(child.GetParameterType(), semanticModel));
+            }
+            else if (HasExplicitNonRouteAttribute(child, semanticModel) || HasSpecialType(child, semanticModel))
+            {
+                continue;
+            }
+            else
+            {
+                resolvedParameterSymbols.Add(child);
+            }
+        }
+        return resolvedParameterSymbols.ToImmutableArray();
+    }
+
+    private static bool HasSpecialType(ISymbol child, SemanticModel semanticModel)
+    {
+        var type = child.GetParameterType() as INamedTypeSymbol;
+        if (type == null)
+        {
+            return false;
+        }
+
+        if (type.IsType(typeof(CancellationToken).FullName!, semanticModel))
+        {
+            return true;
+        }
+
+        if (type.IsType("Microsoft.AspNetCore.Http.HttpContext", semanticModel))
+        {
+            return true;
+        }
+
+        if (type.IsType("Microsoft.AspNetCore.Http.HttpRequest", semanticModel))
+        {
+            return true;
+        }
+
+        if (type.IsType("Microsoft.AspNetCore.Http.HttpResponse", semanticModel))
+        {
+            return true;
+        }
+
+        if (type.IsType("System.Security.Claims.ClaimsPrincipal", semanticModel))
+        {
+            return true;
+        }
+
+        if (type.IsType("Microsoft.AspNetCore.Http.IFormFileCollection", semanticModel))
+        {
+            return true;
+        }
+
+        if (type.IsType("Microsoft.AspNetCore.Http.IFormFile", semanticModel))
+        {
+            return true;
+        }
+
+        if (type.IsType("System.IO.Stream", semanticModel))
+        {
+            return true;
+        }
+
+        if (type.IsType("System.IO.Pipelines.PipeReader", semanticModel))
+        {
+            return true;
+        }
+
+        return false;
+    }
+
+    private static bool HasExplicitNonRouteAttribute(ISymbol child, SemanticModel semanticModel)
+    {
+        var fromBodyMetadataType = semanticModel.Compilation.GetTypeByMetadataName("Microsoft.AspNetCore.Http.Metadata.IFromBodyMetadata");
+        var fromFormMetadataType = semanticModel.Compilation.GetTypeByMetadataName("Microsoft.AspNetCore.Http.Metadata.IFromFormMetadata");
+        var fromHeaderMetadataType = semanticModel.Compilation.GetTypeByMetadataName("Microsoft.AspNetCore.Http.Metadata.IFromHeaderMetadata");
+        var fromQueryMetadataType = semanticModel.Compilation.GetTypeByMetadataName("Microsoft.AspNetCore.Http.Metadata.IFromQueryMetadata");
+        var fromServicesMetadataType = semanticModel.Compilation.GetTypeByMetadataName("Microsoft.AspNetCore.Http.Metadata.IFromServiceMetadata");
```

Comment:
fyi, you may want to consider using a CWT to cache this information for a particular compilation.  that way if this stuff gets called N times (where N may be very high in practice) you only have to compute this once per distinct compilation.  

In practice, due to all the features we have lighting up, and all the code your features run against, you would expect a single compilation to lead to a high number of calls to you.  We cache similar things in embedded languages on roslyn for this reason.

Reply:
It looks like it's already cached in the compilation - https://github.com/dotnet/roslyn/blob/e7c847156fd450f1d5ccea7a00ddf1c9b85e0f25/src/Compilers/Core/Portable/Compilation/Compilation.cs#L1168

Reply:
Note: that only caches 50 items and is shared among all higher layers. So we definitely blow through that :-)

Reply:
Ah, got it.

Do you know of other analyzers using CWT to cache symbols that I could look at?

Alternatively, what about capturing the symbols in a strongly typed class? We do that in other asp.net core analyzers. e.g. https://github.com/dotnet/aspnetcore/blob/main/src/Framework/AspNetCoreAnalyzers/src/Analyzers/RouteHandlers/WellKnownTypes.cs

Reply:
> Alternatively, what about capturing the symbols in a strongly typed class?

could totally do that.  An example of hte CWT are our own detectors.  For example:

```c#

        /// <summary>
        /// Cache so that we can reuse the same <see cref="RegexLanguageDetector"/> when analyzing a particular
        /// compilation model.  This saves the time from having to recreate this for every string literal that features
        /// examine for a particular compilation.
        /// </summary>
        private static readonly ConditionalWeakTable<Compilation, RegexLanguageDetector> _modelToDetector = new();
```

Reply:
Done